simulation/sub8_gazebo/sub8_gazebo_tools/bag_manager.py

```diff
@@ -0,0 +1,148 @@
+
+                if msg is not None:
+                    if (self.nh.get_time() > rospy.Time.from_sec(30)) and (msg_time < self.nh.get_time() - rospy.Duration(30)):
+                        # This fixes a negative time exception if we are within the first 30 seconds of time.
+                        continue
+
+                    if msg_time == last_timestamps[key]:
+                        continue
+                    last_timestamps[key] = msg_time
+
+                    msgs.append([key, (yield msg), (yield msg_time)])
+                else:
+                    print "There's a problem recording {0}".format(key)
+            self.write_to_cache(msgs)
+
+        self.dump()
+
+    @txros.util.cancellableInlineCallbacks
+    def make_dict(self):
+        '''
+        Generate caching dictionary from a yaml file.
+        '''
+        with open(self.diag_dir + 'messages_to_bag.yaml', 'r') as f:
+            messages_to_bag = yaml.load(f)
+
+        # Set bagging parameters
+        self.time_step = messages_to_bag['PARAMS']['time_step']
+        self.pre_cache_time = messages_to_bag['PARAMS']['pre_cache_time']
+        self.post_cache_time = messages_to_bag['PARAMS']['post_cache_time']
+
+        self.cache_dict = {}
+        msgs = messages_to_bag['MESSAGES']
+        for msg in msgs.values():
+            # Get message information
+            msg_topic = msg['message_topic']
+            msg_type = msg['message_type']
+            msg_name = msg['message_name']
+
+            # Import the message
+            exec("from {0}.msg import {1}".format(msg_type, msg_name))
+
+            # Create subscriber and add to dictionary
+            self.cache_dict[msg_topic] = yield self.nh.subscribe(msg_topic, eval(msg_name))
+
+    @txros.util.cancellableInlineCallbacks
+    def dump(self):
+        '''
+        Save cached bags and save the next 'post_cache_time' seconds.
+        '''
+        self.dumping = True
+        # Make bag directory if it doesn't exist
+        directory = os.path.join(self.diag_dir, 'bags')
+        if not os.path.exists(directory):
+            os.makedirs(directory)
+
+        bag = rosbag.Bag(os.path.join(directory, str(int(time.time())) + '.bag'), 'w')
+
+        gen = self.read_from_cache()
+
+        last_timestamps = dict(self.cache_dict)
+
+        print "Saving post-fail data. Do not exit."
+        for i in range(int(self.post_cache_time / self.time_step)):
+            for key in self.cache_dict:
+                msg = self.cache_dict[key].get_last_message()
+                msg_time = self.cache_dict[key].get_last_message_time()
+                if msg is not None:
+                    if msg_time < self.nh.get_time() - rospy.Duration(.5):
+                        # If the message is too old, forget about it.
+                        continue
+
+                    # Lang -- this doesn't work
+                    if msg_time == last_timestamps[key]:
```

Comment:
Yes I think that is right. It just needs to be a dictionary with the same keys as the caching dict so it can add timestamps to that.
It is supposed to dump to a bag here instead of save the messages (I copied this from the start_caching function but obv forgot to change some things).
Try this line instead:
```bag.write(key, (yield msg), t=(yield msg_time))```